### ISSUE TITLE
Userscript support

### ISSUE THREAD

**Issue by kotx** [Vendicated/Vencord#35]
This would be more convenient than installing a browser extension every time Vencord is updated.
Basically the userscript just needs to include
```js
// ==UserScript==
// @name        Vencord
// @namespace   https://github.com/Vendicated/Vencord
// @match       *://*.discord.com/*
// @inject      content
// @run-at      document-start
// @version     1.0
// @homepageURL https://github.com/Vendicated/Vencord
// @supportURL  https://github.com/Vendicated/Vencord/issues
// ==/UserScript==

(0, eval)('<contents of dist/Vencord.js without "use strict">');
```
Note the `match`, `inject`, and `run-at` directives are very similar to the extension's manifest.json keys.

The `"use script";` must be removed because otherwise `Vencord` is not exposed in the global scope.
We cannot stick it in a script element as the browser extension does because it is blocked by CSP. Supposedly we can override this with a `<meta http-equiv="..." content="...">` tag? Unsure.

This also doesn't have autoupdating. We can autoupdate via a URL `@downloadURL <url>` directive. This most likely requires that `@version <version>` be bumped every time the extension updates.

Thoughts about autoupdating:
1. We can host a local server for autoupdates somehow
2. We can host a remote server/GitHub Pages of a built userscript (insecure?)
3. We can just spit out a userscript to paste into the browser when the pnpm build command is run (for now) (aka manual updating)

We also can't have `@downloadURL` as a `file://` URL, it doesn't work.

**Comment by MeguminSama** [comment]
it would be trivial to have a gh worker compile vencord and create a build artifact that can be used for the downloadURL

**Comment by Cynosphere** [comment]
Violentmonkey on Firefox currently has an issue where `Object.defineProperty` on `window` at `@run-at document-start` doesn't work, meaning webpack patching won't work.

**Comment by lol219** [comment]
Itls à good idea

**Comment by Vendicated** [comment]
Added in https://github.com/Vendicated/Vencord/commit/d26196d6c52c092a8e2ef302a2aeec538ff2fe23

Works fine except that Firefox ViolentMonkey bug cyn mentioned. Also runs into csp issues when trying to fetch external urls like github/unpkg

I am not sure how to fix these two issues

**Comment by Vendicated** [comment]
I don't think there's a way to fix the CSP issue without doing a big refactor to use a @grant which means window -> unsafeWindow and it just overcomplicates things. I solved the issue in the proper browser addon by making a background script but this is not possible in UserScripts. The best solution would probably to use an anti csp extension like [this one (Disclaimer: I never used it and have no idea whether it works and is safe)](https://github.com/PhilGrayson/chrome-csp-disable) unless someone else has a better idea / could contribute a solution.
In any case, CSP bypass isn't even that important, only minor features like petpets dynamic library import from cdn don't work so I think it's fine

**Comment by Vendicated** [comment]
Hi Mister grant sorry for pinging you 🚎

**Comment by kiriya-aoi** [comment]
For the purposes of having a version number that increments for auto-updating the userscript, could we use the commit number via e.g. `git describe`?

**Comment by Vendicated** [comment]
Doesn't its updating system use semver?

**Comment by Vendicated** [comment]
Updating should work now since it just uses the current timestamp as version. I don't think csp bypass is doable from userscript, so this issue is basically done now

**Comment by gBasil** [comment]
Aren't there several userscript APIs to circumvent CSP in some cases? `GM_addElement`, `GM_getResourceURL`, and `GM_xmlhttpRequest`.

**Comment by Vendicated** [comment]
please don't necro old resolved issue threads. this issue is more than a year old

yes, those would likely work for some use cases, but that would require us to write specialised UserScript code, which is out of scope, and would still not cover some other use cases. not to mention, tampermonkey already seems to have workarounds for csp issues with which multiple people reported success
